### ISSUE TITLE
Support snake case relationships when these are specified

### ISSUE THREAD

**Issue by JapSeyz** [cloudcreativity/laravel-json-api#315]
Hi,

I was trying to save three belongsTo relationships, but for some reason only the first one would go through; While investing it, I dd'ed here https://github.com/cloudcreativity/laravel-json-api/blob/develop/src/Eloquent/AbstractAdapter.php#L370 and saw that it had fetched the relationship before saving, what's the idea behind this?

Eg. when trying to save a client_id on a user, this is the $record being passed to persist:

....{"client_id":1,"client":{"id":1,"currency":"EUR","name":"client-name"}}.

It seems like a redundant fetch from the database?

I also finally found my issue with only some of the relationships being saved, here: https://github.com/cloudcreativity/laravel-json-api/blob/develop/src/Adapter/AbstractResourceAdapter.php#L205

What's the idea behind Camelizing the method-name, when Laravel famously uses snake-case for relationships?

The package is absolutely wonderful, but I'll see if I can get a PR in with some documentation.

Cheers

**Comment by JapSeyz** [comment]
Especially this paragraph: ["By default this will assume that the Eloquent relation name is the same as the JSON API relation name - author in the example above. If this is not the case, you can provide the Eloquent relation name as the first function argument."](https://laravel-json-api.readthedocs.io/en/latest/basics/adapters/#adapters) threw me for a loop, as it actually formats the method-name and doesn't map it directly, as it states. Would it be possible to get a config-setting for this?

Cheers

**Comment by lindyhopchris** [comment]
Hi!

> What's the idea behind Camelizing the method-name, when Laravel famously uses snake-case for relationships?

Where are you getting that from?

Our relationship methods are always camel-case, as shown in the Laravel docs - e.g. this example:

```php
    /**
     * Get the user's history.
     */
    public function userHistory()
    {
        return $this->hasOneThrough('App\History', 'App\User');
    }
```

Which is in the Laravel Eloquent docs.

**Comment by JapSeyz** [comment]
@lindyhopchris Ah I am sorry, must've been something I've seen somewhere and adapted, and likewise with other developers I've met, as all of the Laravel delopers I meet at meetups also use the snake_case for relationships to keep it similar to attributes.

Either way, I forked this last night and removed the forcing of Camelcase format, and respected the attribute that the user wrote in the resource.

**Comment by lindyhopchris** [comment]
Are your method names snake cased then?

**Comment by lindyhopchris** [comment]
As in relationship method names, i.e. would you implement the above example as:

```php
    /**
     * Get the user's history.
     */
    public function user_history()
    {
        return $this->hasOneThrough('App\History', 'App\User');
    }
```

**Comment by JapSeyz** [comment]
Method names for relationships are snake_cased, regular methods are camelCased.

**Comment by lindyhopchris** [comment]
So we've never done that because it's not compliant with PSR1:

> Method names MUST be declared in camelCase().
https://www.php-fig.org/psr/psr-1/

Saying that, if you submit a PR that allows underscore, but by default assumes the PSR1 standard of camel casing, and has tests, I'd definitely consider merging it.

**Comment by JapSeyz** [comment]
I'll see if I can whip something up in the weekend.

I know of a few other Laravel developers who also use relationships as snake_case, I am a bit curious as to where that trend came from now. Probably to match with attributes when having an object instance, like `$user->user_history->location_id`, rather than `$user->userHistory->location_id`, which looks odd because of mixed casing.

**Comment by JapSeyz** [comment]
Awesome.

Just for my own curiosity, do you then have $user->userRole->role_name usually, or do you also camelCase the sql columns?

Cheers

**Comment by lindyhopchris** [comment]
We snake case column names as that seems to be the database convention. It doesn't both us that one is snake cased and the other camel cased - because the two things are not the same. An attribute that is a value on a column in a database is completely different from a related model, so we've never felt the need to enforce them to be consistent with each other.

I'd also suspect the majority of people follow the camel case convention otherwise this issue would have come up a long time ago! The camel case conversion has been in this package for a significant amount of time.

But as I say, happy to accommodate. The default should be to camel case but totally accept that if the developer provided `user_history` as the Eloquent relationship name, that should be respected!

**Comment by JapSeyz** [comment]
Alright, I have made the changes in my own fork, which basically checks the user-provided relationship name, and if it's not found it camelcases it and uses that.

It becomes more difficult with the include, as it's being force-camelized, albeit I want include="user.user_history" to include user_history, and you'd like userHistory.

I'd need to dive deeper into the code to see if it's possible to get the relationship name at the point in the code where it camelizes that, as I don't think any resources/models has been instantiated yet.

I'll validate in the weekend and see what I can do.

**Comment by lindyhopchris** [comment]
You can override include paths already, using the `$includePaths` property on the adapter. Looking at the code, this will be respected rather than camelized - so shouldn't require any changes. The relevant line is here:
https://github.com/cloudcreativity/laravel-json-api/blob/develop/src/Eloquent/Concerns/IncludesModels.php#L144

It's a little bit clunky having to define the relationship mapping in two places, and I'm already aware it would be better for the include path to fall back to the relationship method on the adapter. However, I think that's a separate issue so shouldn't be in scope of this fix. It's also at that point becoming a potentially breaking feature rather than a bug fix.

So the PR should fix the Eloquent relationship name being camel cased if provided by the developer - i.e. it should use exactly what the developer provides rather than camel casing it. As far as I can see the PR will not need to change anything around include paths as these can already be specified exactly by the developer.

**Comment by JapSeyz** [comment]
It does seem a bit wonky having to define it in two places.

As far as I can tell, the line that is causing issues is this: https://github.com/cloudcreativity/laravel-json-api/blob/develop/src/Eloquent/Concerns/IncludesModels.php#L148 if no ´$includePaths´ has been defined.

The converted (camelized) includePath is being used as the Eloquent Relationship through here: https://github.com/cloudcreativity/laravel-json-api/blob/develop/src/Eloquent/Concerns/IncludesModels.php#L118

 and then https://github.com/cloudcreativity/laravel-json-api/blob/develop/src/Eloquent/Concerns/IncludesModels.php#L91 in the QueryBuilder ->with()

I am not sure I can "fix" the eloquent_relationship name without messing with the includePath. Perhaps a config value to pick camelized/snake_cased relationship naming would be a fix?

**Comment by lindyhopchris** [comment]
Sorry, but I don't really understand why L148 is a problem, when L144 will return first if there is defined include path.

**Comment by lindyhopchris** [comment]
I.e. if you don't want it to camelize, the current code provides you a path to do that.

I totally appreciate the duplication is not ideal, but this needs to be a bug fix, not changing the current behaviour.

The current code provides you a way out if you don't want to use camel case - i.e. define the include path yourself. 

The bug that needs to be fixed is the relationship methods don't provide you a way out because it camel cases the name you've provided rather than just using the name you've provided.

Removing the duplication is something completely separate, that probably needs to be done at a major release as it'll most likely be breaking.

**Comment by JapSeyz** [comment]
Line 148 was an issue as I assumed you wanted rid of the include_path altogether.

I am not entirely sure what you mean with "The bug that needs to be fixed is the relationship methods don't provide you a way out because it camel cases the name you've provided rather than just using the name you've provided.". To fix this, I would exactly need to change Line 148, unless you mean it's camelizing it in a different location?

I think I'll stick to my fork for now, as I have made some changes in other places as well, which allows for, for-example multiple pivots relations between the same two resources. Again something I specifically need and not something that seems to have been an issue before, so I'll keep it separate in my own fork.

**Comment by lindyhopchris** [comment]
Ah ok, well my point is that at the moment you can get round the include problem by simply adding this to the adapter:

```php
protected $includePaths = [
  'user_history' => 'user_history',
];
```

Because that won't be camelized. Not saying at all that it's perfect, but it's definitely an acceptable workaround. Line 148 won't be hit with the above config.

**Comment by JapSeyz** [comment]
Thats' fine, I recognize that the method will return before 148 if there's an includesPaths set, however I do have quite a few relationships lined up, so it'd be a lot of unnecessary bulk having to defined them all twice.

A config value would perhaps enable us to return the user-sent value at line 146 if it's true, and return to line 148 if it's false.

It's essentially a toggle for the Camelization, would that be of interest?

**Comment by lindyhopchris** [comment]
On the `DeserializesAttributes` trait there is a method `modelKeyForField($field, $model)`. This is doing the conversion from a JSON API field name to a model column name.

What we need is an equivalent for working out the model relation (i.e. the function on the model for a relation).

I think therefore the `IncludesModels` needs to be modified as follows:

```php
protected $camelCaseRelations = true;

// ...

protected function modelRelationForField($field)
{
   return $this->camelCaseRelations ? Str::camelize($field) : $field;
}
```

Then replace the other instances of `Str::camelize` with `$this->modelRelationForField()`

Then the developer:

- Does nothing to keep using camel case
- Changes `$camelCaseRelations` to false if no conversion from the JSON API field name is required (e.g. if you are using snake case in JSON API and snake case on your model).
- Overloads `modelRelationForField` if they are doing anything else (e.g. using dash case in JSON API and underscore on their model).

**Comment by lindyhopchris** [comment]
I've just realised that this line:
https://github.com/cloudcreativity/laravel-json-api/blob/develop/src/Adapter/AbstractResourceAdapter.php#L205

Does not need changing. It is working out the method name on the adapter, not the method name on the model. My view is that it is completely reasonable for us to be following PSR1 on our own adapter. However it is provided as a method so that the developer can overload if they want. So if a developer does not want to follow PSR1 on their JSON API adapter, then that method should be overloaded.

**Comment by lindyhopchris** [comment]
The bit that needs changing for the relationship names to be underscored is as follows. Once the change to the trait suggested above has been made, the private guess relation method can be changed from:

```php
    /**
     * @return string
     */
    private function guessRelation()
    {
        list($one, $two, $caller) = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 3);

        return $caller['function'];
    }
```

to:

```php

   protected function modelRelationForMethod($method)
   {
       return $this->camelCaseRelations ? $method : Str::underscore($method);
   }

    private function guessRelation()
    {
        list($one, $two, $caller) = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 3);

        return $this->modelRelationForMethod($caller['function']);
    }
```

**Comment by lindyhopchris** [comment]
So `modelRelationForField` converts JSON API field to the model relation method. `modelRelationForMethod` converts the adapter's method name to the model's method name.

**Comment by lindyhopchris** [comment]
I'm happy to make these changes if you think they will meet your use case?

**Comment by JapSeyz** [comment]
My take on 
```php
 protected function methodForRelation($field)
    {
        $method = Str::camelize($field);
        return method_exists($this, $method) ? $method : null;
    }
```

would be:

```php
 protected function methodForRelation($method)
    {
        return method_exists($this, $method) ? $method : Str::camelize($method);
    }
```

To always favour what the developer provides first, and then fallback onto the package specific setup.

I'll have a look at the rest of the comments shortly

**Comment by lindyhopchris** [comment]
That's not what that method is doing though. `$field` is the JSON API field name. For example, `user-history`. So PSR1 says method names should be camelized, so the correct default behaviour is to camelize the JSON API field name.

The string it is returning refers to the method name on the adapter, not the model - which is why that method is on the `AbstractResourceAdapter`, which the Eloquent adapter extends.

**Comment by JapSeyz** [comment]
I think we might be talking past each other.

My case is to get rid of the forced Camelization in both resources and models, not having it Camelized in the resource and snaked in the model.

Cheers

**Comment by lindyhopchris** [comment]
So as an example, if you do `GET /api/user/1/user-history` the `users` adapter is asked for relation `user-history`. 

So the abstract adapter calls `methodForRelation('user-history')` and gets `userHistory` back. It then calls `$this->userHistory()`.

Within `userHistory()` you then call `$this->hasMany('user_history')` which tells the relationship object that the model key is `user_history`.

**Comment by lindyhopchris** [comment]
Ok so if that's the case, the proposed adjustments above will meet your requirements. I'll push a branch for you to try.

**Comment by lindyhopchris** [comment]
Note for self, think this would do it:

```php
protected function methodForRelation($field)
{
    if (method_exists($this, $field)) {
        return $field;
    }

    $method = Str:camelize($field);

    return method_exists($this, $method) ? $method : null;
}
```

**Comment by JapSeyz** [comment]
Sounds great, I'll test it once the branch is ready.

Cheers

**Comment by lindyhopchris** [comment]
@JapSeyz 

```bash
$ composer require cloudcreativity/laravel-json-api:dev-issue315`
```

**Comment by JapSeyz** [comment]
Sweet, I'll test it out and get back to you

**Comment by JapSeyz** [comment]
I don't know if this is a bug, but outside of the test-branch, ( I haven't had the time to test it yet), when I submit data for creating a new record, with a belongs_to relationship, it doesn't attempt to insert `car_id` into the db, but rather `car`.

Eg. creating a person, and then creating a car that belongs to that user. When running the query with {type: 'users', id: 1}, it creates a SQL statement trying to insert user: 1, rather than user_id: 1.


Have you experienced this before?

Cheers

**Comment by JapSeyz** [comment]
The issue315 branch seems to allow snake_cased relations 👍

**Comment by lindyhopchris** [comment]
Great, I'll get that merged into `develop` then.

> I don't know if this is a bug, but outside of the test-branch, ( I haven't had the time to test it yet), when I submit data for creating a new record, with a belongs_to relationship, it doesn't attempt to insert car_id into the db, but rather car.

The relationship solution has been in use for a relatively long time, and in multiple production apps that are extensively tested. I've never seen this, so my gut feel is it's something you're doing. If the problem persists, I'd suggest creating a separate issue and would need to see a lot more info on what you're actually doing! FYI the belongs to relationship uses `$relation->associate($model)` to set the model, rather than trying to change the (e.g.) `user_id` column value directly.
